### PR TITLE
Reorganization and upgrade of spikekill.php

### LINKED ISSUE #3636: Spikekill method variable is misinterpreted when calculating overall statistics
# Describe the bug

The Spikekill script accepts a few input parameters when it's run, one of which is the "method".  This is either `stddev`, `variance`, `float`, or `fill`.  At the beginning of the script, these four values are assigned the numbers 1 through 4, respectively.  During the `calculateOverallStatistics` function, the values of 1 and 2 are misinterpreted such that a value of 1 indicates variance, and 2 indicates stddev.

## To Reproduce

This behavior occurs every time the spikekill script is run.

## Expected behavior

If the variables are interpreted correctly, this will result in the DryRun table containing accurate data.

## Screenshots

n/a

## Desktop (please complete the following information)

n/a

## Smartphone (please complete the following information)

n/a

## Additional context

n/a

### LINKED ISSUE #3675: Spikekill fills in NaN values outside specified time range
# Describe the bug

When running Spikekill using GapFill method on a graph containing holes in the data, Spikekill will fill in values that are outside the time range.

## To Reproduce

1. Find a graph with holes in the data.  Here is my example graph.  (Red arch line added.)
image

2. Zoom into the graph at a time range that contains a hole, but not the first several holes.  In the example above, I zoomed into the range where I drew the red arch.

3. Run Spikekill using the GapFill method.

4. Observe fixed graph.  This looks good so far because the hole was filled in and the spike was mitigated.
image

5. Zoom out of the 'fixed' graph and observe what happened.  I drew the red arch once again for you to see that my hole is filled in and the spike is mitigated.  However, there are many holes in the left part of the graph that are suddenly filled in, despite being outside the time range that was specified.
image

6. Since I had Cacti's debug level set to Debug, I can observe in the log file that the time range is limited to the range of the graph above (which is perfect!).  I can also observe in the log that Spikekill mitigated 14 spikes, which is far more than existed in my time range.
```
2020/07/09 14:24:21 - SPIKEKILL CactiUser:admin, File:ping_int_velocity_tunb_ping_1282.rrd, Method:GapFill, OutStart:1594060746, OutEnd:1594061254, AvgNan:last
2020/07/09 14:24:22 - SPIKE DEBUG: number of RRAs: 12
2020/07/09 14:24:22 - SPIKE DEBUG: number of DSes: 1
2020/07/09 14:24:22 - SPIKE DEBUG: Limited to window: Jul 6, 2020 13:39:06 thru Jul 6, 2020 13:47:34
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7600000000e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.8545000000e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.8830000000e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7500000000e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7570000000e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue 5.9415000000e+01 with last value 2.7636666667e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7636666667e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue 5.9415000000e+01 with last value 2.7636666667e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7636666667e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue 5.9415000000e+01 with last value 2.7636666667e+01
2020/07/09 14:24:22 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7636666667e+01
2020/07/09 14:24:23 - SPIKEKILL DEBUG: replacing dsvalue 5.9415000000e+01 with last value 2.7800000000e+01
2020/07/09 14:24:23 - SPIKEKILL DEBUG: replacing dsvalue 5.9415000000e+01 with last value 2.7636666667e+01
2020/07/09 14:24:23 - SPIKEKILL DEBUG: replacing dsvalue NaN with last value 2.7636666667e+01
2020/07/09 14:24:24 - WEBUI WARNING: Removed '14' Spikes from '/var/www/html/cacti/rra/ping_int_velocity_tunb_ping_1282.rrd', Method:'4'
```

## Expected behavior

I would expect that the time range provided to Spikekill would prevent any modifications of data outside the given time range.

## Screenshots

Provided above.

## Desktop (please complete the following information)

n/a

## Smartphone (please complete the following information)

n/a

## Additional context

None

### PR DESCRIPTION
**Pull request by riversdev0**
Fixes #3636
Fixes #3675
Implements #3620
Lays foundation for #3724
Lays foundation for #3725

### CODE REVIEW

**Comment by riversdev0** [comment]
I obviously did something unwise to cause a bunch of conflicts.  Not sure what to do from here, so please provide guidance regarding how i can resolve the conflicts.  This "develop" version of the Spikekill script can safely overwrite, in entirety, the "1.2.x" version.

File: lib/spikekill.php [hunk 1]
```diff
@@ -181,61 +210,27 @@ private function initialize_spikekill() {
...
+		   correct values from the default (which came from the user's settings,
+		   or the global settings, or statically defined by the class */
+		if ($this->avgnan == '') $this->avgnan = $uavgnan;
+		if ($this->method == '') $this->method = $umethod;
+		if ($this->numspike == '') $this->numspike = $unumspike;
+		if ($this->stddev == '') $this->stddev = $ustddev;
+		if ($this->percent == '') $this->percent = $upercent;
+		if ($this->outliers == '') $this->outliers = $uoutliers;
+		if ($this->dsfilter == '') $this->dsfilter = $udsfilter;
+		if ($this->absmax == '') $this->absmax = $uabsmax;
```
Comment (TheWitness) [inline lib/spikekill.php:233]: @riversdev0, we always require curly braces around IF/THEN/ELSE.  Can you update to match that style.

Comment (riversdev0) [reply lib/spikekill.php]: Sure, done.

**Comment by netniV** [comment]
I would update this to resolve the conflicts but it would appear that the repository is missing (showing as unknown above by github).

**Comment by riversdev0** [comment]
Because I don't really understand all of the different aspects of Github's workflow yet, I've seen some conflicts and other things that I thought were my fault, and I've done some not-so-wise things in an attempt to resolve stuff.  In this particular instance, I deleted the repository and re-forked it fresh because I thought that would clear something up, but I just made it worse.  So, not sure what to do from here.  I could submit a new PR with the same changes, would that help?

**Comment by netniV** [comment]
I will make a commit, please check it and then post any fixes as a new PR 👍 

Always use this method of creating patches based on your own fork:

1. Update the core branch (ie, develop or 1.2.x)
2. Create a new branch from the updated core
3. Do your mods
4. Commit
5. Push
6. Create a PR
7. Once committed to our core, delete the branch

You can at this point update your core branch again so you are fully updated.  If you have other branches based of your local core, you can then use rebase or merge depending on your preference.  I tend to use rebase as all my branch changes are local to that branch so ensure my changes aren't breaking things.

Hopefully this explains things a bit.

**Comment by netniV** [comment]
The above patch has now been pushed.

**Pull request closed without merging**